File: recorder/src/utils/run-layout.js
/**
 * run-layout.js - 单次录制 run_* 目录布局（唯一路径来源）
 *
 * 布局：
 *   run_<timestamp>/
 *     meta.json                 # 录制元信息，也是下游翻译入口锚点
 *     record/                   # 录制原始数据
 *       actions/
 *       snapshots/
 *       screenshots/            # 可选
 *       recorder.log
 */

import fs from 'fs';
import path from 'path';

/** 录制元信息文件名（位于 run 根目录，翻译入口锚点） */
export const META_FILENAME = 'meta.json';

/** 截图子目录名（位于 record/ 下） */
export const SCREENSHOTS_SUBDIR = 'screenshots';

/** 录制数据根目录名（相对 runDir） */
export const RUN_RECORD_SUBDIR = 'record';

/** 录制：actions 子目录（相对 runDir） */
export const RECORD_ACTIONS_REL = `${RUN_RECORD_SUBDIR}/actions`;

/** 录制：snapshots 子目录（相对 runDir） */
export const RECORD_SNAPSHOTS_REL = `${RUN_RECORD_SUBDIR}/snapshots`;

/** 录制：截图子目录（相对 runDir） */
export const RECORD_SCREENSHOTS_REL = `${RUN_RECORD_SUBDIR}/${SCREENSHOTS_SUBDIR}`;

/** 录制：日志文件（相对 runDir） */
export const RECORD_LOG_REL = `${RUN_RECORD_SUBDIR}/recorder.log`;

/**
 * @param {string} runDir
 * @returns {string}
 */
export function getMetaPath(runDir) {
  return path.join(runDir, META_FILENAME);
}

/**
 * 录制阶段路径
 *
 * @param {string} runDir
 * @returns {{
 *   recordDir: string,
 *   actionsDir: string,
 *   snapshotsDir: string,
 *   screenshotsDir: string,
 *   recorderLog: string,
 * }}
 */
export function getRecordPaths(runDir) {
  const recordDir = path.join(runDir, RUN_RECORD_SUBDIR);
  return {
    recordDir,
    actionsDir: path.join(recordDir, 'actions'),
    snapshotsDir: path.join(recordDir, 'snapshots'),
    screenshotsDir: path.join(recordDir, SCREENSHOTS_SUBDIR),
    recorderLog: path.join(recordDir, 'recorder.log'),
  };
}

/**
 * 创建录制目录结构（record/actions、record/snapshots 等）
 *
 * @param {string} runDir
 * @param {{ screenshots?: boolean }} [options]
 */
export function ensureRecordLayout(runDir, options = {}) {
  const { screenshots = false } = options;
  const { actionsDir, snapshotsDir, screenshotsDir, recordDir } = getRecordPaths(runDir);
  fs.mkdirSync(recordDir, { recursive: true });
  fs.mkdirSync(actionsDir, { recursive: true });
  fs.mkdirSync(snapshotsDir, { recursive: true });
  if (screenshots) {
    fs.mkdirSync(screenshotsDir, { recursive: true });
  }
}

/**
 * Dashboard / API 可预览的相对路径（相对 runDir）
 */
export const DASHBOARD_PREVIEW_REL_PATHS = [
  META_FILENAME,
  RECORD_LOG_REL,
];

File: recorder/translate_gui/internal/translator/run_layout.go
package translator

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// ==================== 目录名常量 ====================

const RunRecordSubdir = "record"
const RunTranslateSubdir = "translate"
const ScreenshotsSubdir = "screenshots"

// ==================== 相对路径（相对 runDir） ====================

const RecordActionsRel = RunRecordSubdir + "/actions"
const RecordSnapshotsRel = RunRecordSubdir + "/snapshots"
const RecordLogRel = RunRecordSubdir + "/recorder.log"

const TranslateGenerateLogRel = RunTranslateSubdir + "/logs/generate.log"
const TranslatePreprocessRel = RunTranslateSubdir + "/preprocess"
const TranslatePhase1Rel = RunTranslateSubdir + "/phase1"
const TranslatePhase1StepsJsonRel = TranslatePhase1Rel + "/structured_steps.json"
const TranslatePhase1StepsXmlRel = TranslatePhase1Rel + "/structured_steps.xml"
const TranslatePhase1LlmRawXmlRel = TranslatePhase1Rel + "/llm_raw_batches.xml"
const TranslatePhase1ErrorsJsonRel = TranslatePhase1Rel + "/errors.json"
const TranslatePhase2SlicesJsonRel = RunTranslateSubdir + "/phase2/case_slices.json"
const TranslatePhase2CoverageMdRel = RunTranslateSubdir + "/phase2/coverage.md"
const TranslatePhase3AgentsTxtRel = RunTranslateSubdir + "/phase3/agents.txt"
const TranslatePhase4CasesMdRel = RunTranslateSubdir + "/phase4/cases.md"
const TranslateLlmAuditRel = RunTranslateSubdir + "/llm_audit"

// ==================== 路径获取 ====================

func getMetaPath(runDir string) string {
	return filepath.Join(runDir, MetaFilename)
}

func readTargetURLFromMeta(runDir string) string {
	data, err := os.ReadFile(getMetaPath(runDir))
	if err != nil {
		return ""
	}
	var meta Meta
	if err := json.Unmarshal(data, &meta); err != nil {
		return ""
	}
	if initialURL := strings.TrimSpace(meta.InitialURL); initialURL != "" {
		return initialURL
	}
	return strings.TrimSpace(meta.TargetURL)
}

type RecordPaths struct {
	RecordDir      string
	ActionsDir     string
	SnapshotsDir   string
	ScreenshotsDir string
	RecorderLog    string
}

func getRecordPaths(runDir string) RecordPaths {
	recordDir := filepath.Join(runDir, RunRecordSubdir)
	return RecordPaths{
		RecordDir:      recordDir,
		ActionsDir:     filepath.Join(recordDir, "actions"),
		SnapshotsDir:   filepath.Join(recordDir, "snapshots"),
		ScreenshotsDir: filepath.Join(recordDir, ScreenshotsSubdir),
		RecorderLog:    filepath.Join(recordDir, "recorder.log"),
	}
}

type TranslatePaths struct {
	TranslateDir        string
	GenerateLog         string
	PreprocessDir       string
	DiffsDir            string
	EnrichedDir         string
	MergedDir           string
	Phase1Dir           string
	StructuredStepsJson string
	StructuredStepsXml  string
	LlmRawXml           string
	ErrorsJson          string
	Phase2Dir           string
	CaseSlicesJson      string
	CasesCoverageMd     string
	Phase3Dir           string
	AgentsTxt           string
	Phase4Dir           string
	CasesMd             string
	LlmAuditDir         string
}

func getTranslatePaths(runDir string) TranslatePaths {
	translateDir := filepath.Join(runDir, RunTranslateSubdir)
	preprocessDir := filepath.Join(translateDir, "preprocess")
	phase1Dir := filepath.Join(translateDir, "phase1")
	phase2Dir := filepath.Join(translateDir, "phase2")
	phase3Dir := filepath.Join(translateDir, "phase3")
	phase4Dir := filepath.Join(translateDir, "phase4")
	return TranslatePaths{
		TranslateDir:        translateDir,
		GenerateLog:         filepath.Join(translateDir, "logs", "generate.log"),
		PreprocessDir:       preprocessDir,
		DiffsDir:            filepath.Join(preprocessDir, "diffs"),
		EnrichedDir:         filepath.Join(preprocessDir, "enriched"),
		MergedDir:           filepath.Join(preprocessDir, "merged"),
		Phase1Dir:           phase1Dir,
		StructuredStepsJson: filepath.Join(phase1Dir, "structured_steps.json"),
		StructuredStepsXml:  filepath.Join(phase1Dir, "structured_steps.xml"),
		LlmRawXml:           filepath.Join(phase1Dir, "llm_raw_batches.xml"),
		ErrorsJson:          filepath.Join(phase1Dir, "errors.json"),
		Phase2Dir:           phase2Dir,
		CaseSlicesJson:      filepath.Join(phase2Dir, "case_slices.json"),
		CasesCoverageMd:     filepath.Join(phase2Dir, "coverage.md"),
		Phase3Dir:           phase3Dir,
		AgentsTxt:           filepath.Join(phase3Dir, "agents.txt"),
		Phase4Dir:           phase4Dir,
		CasesMd:             filepath.Join(phase4Dir, "cases.md"),
		LlmAuditDir:         filepath.Join(translateDir, "llm_audit"),
	}
}

func ensureTranslateLayout(runDir string) {
	p := getTranslatePaths(runDir)
	os.MkdirAll(filepath.Join(p.TranslateDir, "logs"), 0755)
	os.MkdirAll(p.DiffsDir, 0755)
	os.MkdirAll(p.EnrichedDir, 0755)
	os.MkdirAll(p.MergedDir, 0755)
	os.MkdirAll(p.Phase1Dir, 0755)
	os.MkdirAll(p.Phase2Dir, 0755)
	os.MkdirAll(p.Phase3Dir, 0755)
	os.MkdirAll(p.Phase4Dir, 0755)
	os.MkdirAll(p.LlmAuditDir, 0755)
}
